File: Assets/MLRepo/ml.72201f1e-ba18-487e-adc4-4381ae10e21b.davidmatos_doorbutton/DoorButton/Scripts/DoorButtonBase.cs
using ML.SDK;
using UnityEngine;

    public class DoorButtonBase : MonoBehaviour
    {
        [Header("Door References")]
        [SerializeField] private GameObject doorObject;
        [SerializeField] private GameObject doorOpenedPosition;
        [SerializeField] private GameObject doorClosedPosition;
        [SerializeField] private float doorSpeed = 1f;

        [Header("Audio Settings")]
        [SerializeField] private AudioClip openSound;
        [SerializeField] private AudioClip closeSound;

        // runtime state
        private bool localDoorState = false; // false == closed
        private AudioSource audioSource;

        // network event ids / tokens
        private const string EVENT_ID_TOGGLE = "DoorToggle";
        private const string EVENT_ID_LATE_JOINER = "Door_LateJoiner";
        private EventToken tokenToggle;
        private EventToken tokenLateJoiner;

        void Start()
        {
            audioSource = GetComponent<AudioSource>();
            if (audioSource == null) audioSource = gameObject.AddComponent<AudioSource>();

            // place door in closed position initially
            if (doorObject != null && doorClosedPosition != null)
                doorObject.transform.position = doorClosedPosition.transform.position;

            // register network handlers
            tokenToggle = this.AddEventHandler(EVENT_ID_TOGGLE, OnNetworkDoorToggle);
            tokenLateJoiner = this.AddEventHandler(EVENT_ID_LATE_JOINER, OnNetworkLateJoiner);

            // listen for players joining so master can sync late joiners
            MassiveLoopRoom.OnPlayerJoined += PlayerJoined;
        }

        void OnDestroy()
        {
            // tidy up
            MassiveLoopRoom.OnPlayerJoined -= PlayerJoined;
            // (If your SDK exposes a RemoveEventHandler by token, call it here.
            // sample code left handlers registered in many examples; if you have RemoveEventHandler, use it.)
        }

        void Update()
        {
            UpdateDoorState();
        }

        private void UpdateDoorState()
        {
            if (doorObject == null || doorOpenedPosition == null || doorClosedPosition == null) return;

            Vector3 target = localDoorState ? doorOpenedPosition.transform.position : doorClosedPosition.transform.position;
            doorObject.transform.position = Vector3.Lerp(doorObject.transform.position, target, Time.deltaTime * doorSpeed);
        }

        // Public API: called by local button / clickable. MLClickable passes MLPlayer, many examples accept MLPlayer.
        // Provide both overloads so either wiring style works.
        public void ToggleDoor()
        {
            DoToggle();
        }

        public void ToggleDoor(MLPlayer player)
        {
            DoToggle();
        }

        // internal toggle that enforces master-client authority
        private void DoToggle()
        {
            // Only master client should authoritatively toggle
      //      if (!MassiveLoopClient.IsMasterClient) return;

            bool newState = !localDoorState;
            // apply locally on master immediately (no need to wait for network callback)
            SetDoorState(newState);

            // broadcast to other clients (send int 0/1 + object name so multiple doors won't conflict)
            int stateInt = newState ? 1 : 0;
            this.InvokeNetwork(EVENT_ID_TOGGLE, EventTarget.Others, null, stateInt, this.gameObject.name);
        }

        // Received on other clients (master used EventTarget.Others so master won't re-apply)
        private void OnNetworkDoorToggle(object[] args)
        {
            // args: [int stateInt, string objectName]
            if (args == null || args.Length < 2) return;

            string objName = (string)args[1];
            if (this.gameObject.name != objName) return;

            int stateInt = (int)args[0];
            bool newState = stateInt == 1;
            SetDoorState(newState);
        }

        // Master sends this to sync late joiners
        private void PlayerJoined(MLPlayer player)
        {
            if (!MassiveLoopClient.IsMasterClient) return;

            // send current state to all (or you could target only the new player if API supports)
            int stateInt = localDoorState ? 1 : 0;
            this.InvokeNetwork(EVENT_ID_LATE_JOINER, EventTarget.All, null, stateInt, this.gameObject.name);
        }

        // Handler for late-join sync
        private void OnNetworkLateJoiner(object[] args)
        {
            // args: [int stateInt, string objectName]
            if (args == null || args.Length < 2) return;

            string objName = (string)args[1];
            if (this.gameObject.name != objName) return;

            int stateInt = (int)args[0];
            bool newState = stateInt == 1;
            SetDoorState(newState);
        }

        private void SetDoorState(bool newState)
        {
            if (localDoorState == newState) return;

            localDoorState = newState;

            // play sound (master and non-master will play when they apply)
            if (audioSource != null)
            {
                AudioClip clip = newState ? openSound : closeSound;
                if (clip != null) audioSource.PlayOneShot(clip);
            }
        }
    }

File: Assets/MLRepo/ml.72201f1e-ba18-487e-adc4-4381ae10e21b.davidmatos_doorbutton/DoorButton/Scripts/DoorButtonBase.lua
Namespace("davidmatos_doorbutton");
do -- script DoorButtonBase 
	
	-- get reference to the script
	local DoorButtonBase = LUA.script;
	
	--|||||||||||||||||||||||||||||||||||||||||||||| PUBLIC VARIABLES ||||||||||||||||||||||||||||||||||||||||||||||
	--|||||||||||||||||||||||||||||||||||||||||||||| PUBLIC VARIABLES ||||||||||||||||||||||||||||||||||||||||||||||
	--|||||||||||||||||||||||||||||||||||||||||||||| PUBLIC VARIABLES ||||||||||||||||||||||||||||||||||||||||||||||

    local doorObject = SerializedField("Door Object", GameObject);
    local doorOpenedPosition = SerializedField("Open Position", GameObject);
    local doorClosedPosition = SerializedField("Closed Position", GameObject);
    local doorSpeed = SerializedField("Speed", Number);

    local audioClip_openSound = SerializedField("Open Sound", AudioClip);
    local audioClip_closeSound = SerializedField("Close Sound", AudioClip);

    --|||||||||||||||||||||||||||||||||||||||||||||| SERVER VARIABLES ||||||||||||||||||||||||||||||||||||||||||||||
	--|||||||||||||||||||||||||||||||||||||||||||||| SERVER VARIABLES ||||||||||||||||||||||||||||||||||||||||||||||
	--|||||||||||||||||||||||||||||||||||||||||||||| SERVER VARIABLES ||||||||||||||||||||||||||||||||||||||||||||||

    local serverDoorState = SyncVar(DoorButtonBase, "doorState");

    --|||||||||||||||||||||||||||||||||||||||||||||| PRIVATE VARIABLES ||||||||||||||||||||||||||||||||||||||||||||||
	--|||||||||||||||||||||||||||||||||||||||||||||| PRIVATE VARIABLES ||||||||||||||||||||||||||||||||||||||||||||||
	--|||||||||||||||||||||||||||||||||||||||||||||| PRIVATE VARIABLES ||||||||||||||||||||||||||||||||||||||||||||||

    local localDoorState = false;
    local localPlayer = nil;
    local audioSource = nil;

    --|||||||||||||||||||||||||||||||||||||||||||||| MAIN LOGIC ||||||||||||||||||||||||||||||||||||||||||||||
	--|||||||||||||||||||||||||||||||||||||||||||||| MAIN LOGIC ||||||||||||||||||||||||||||||||||||||||||||||
	--|||||||||||||||||||||||||||||||||||||||||||||| MAIN LOGIC ||||||||||||||||||||||||||||||||||||||||||||||

    local function UpdateDoorState()
        if(localDoorState == true) then
            doorObject.transform.position = Vector3.Lerp(doorObject.transform.position, doorOpenedPosition.transform.position, Time.deltaTime * doorSpeed);
        else
            doorObject.transform.position = Vector3.Lerp(doorObject.transform.position, doorClosedPosition.transform.position, Time.deltaTime * doorSpeed);
        end
    end

    local function ToggleDoor()
        if(localPlayer.isMasterClient == true) then
            localDoorState = not localDoorState;
            serverDoorState.SyncSet(localDoorState);
        end
    end

    local function serverDoorState_OnVariableSet(value)
        localDoorState = value;

        if(localDoorState == true) then
            audioSource.PlayOneShot(audioClip_openSound);
        else
            audioSource.PlayOneShot(audioClip_closeSound);
        end
    end

    --|||||||||||||||||||||||||||||||||||||||||||||| UNITY FUNCTIONS ||||||||||||||||||||||||||||||||||||||||||||||
	--|||||||||||||||||||||||||||||||||||||||||||||| UNITY FUNCTIONS ||||||||||||||||||||||||||||||||||||||||||||||
	--|||||||||||||||||||||||||||||||||||||||||||||| UNITY FUNCTIONS ||||||||||||||||||||||||||||||||||||||||||||||

    function DoorButtonBase.Start()
        localPlayer = Room.GetLocalPlayer();

        audioSource = DoorButtonBase.gameObject.GetComponent(AudioSource);

        serverDoorState.OnVariableSet.Add(serverDoorState_OnVariableSet);

        LuaEvents.AddLocal(DoorButtonBase, "ToggleDoor", ToggleDoor);

        if(SyncVar.Exists(DoorButtonBase, "doorState") == true) then
            localDoorState = serverDoorState.SyncGet();
        end
    end

    function DoorButtonBase.Update()
        UpdateDoorState();
    end
end